### ISSUE TITLE
Support SystemD service configuration

### ISSUE THREAD

**Issue by cloudguru-bb** [juju-solutions/layer-docker#21]
What is the best way of adding docker daemon flags for layer-docker based charms ?

DOCKER_OPTS  allows for the --insecure-registry parameter.  However, most private docker registries that use the --insecure-registry flag often also require flags for -e (email) -u (user name) -p (password) for a users with private registry permissions. Without setting these flags images can't be pulled by the charm.

**Comment by chuckbutler** [comment]
These flags can be passed in using the same method you would pass the --insecure-registry option

I believe that adding the options as so would work:

opts = DockerOpts()
opts.add('e', 'user@host.com')
opts.add('u', 'joe')
opts.add('p', 'bagodonuts')

these will be output in the opts format when calling opts.to_s()

"-e user@host.com -u joe -p bagodonuts" 

As the layer-docker only provides the plumbing, your layer built on top of layer-docker is free to set these fields, and/or could be added to the charm configuration to set these private registry variables.

Ideally, i'd like to see this data get set via relation to a private registry charm, but I also understand that users are settled into third party registries and this should be given more attention. I'll see about getting this in the roadmap once we've stabilized the kubernetes and swarm layers we're actively focusing on.

**Comment by chuckbutler** [comment]
Really good insight @cloudguru-bb, thanks for keeping this issue up to date. As we track the Xenial release, and update layer-docker to work on systemd based systems this will get implemented.

I like the idea of having an intermediary step where we add an ENVFILE mapping to the systemd job to point at /etc/defaults/docker. However I'm open to hear what user's feel is a better way to do this. If we need to fully update the opts parser to point at a systemd job when on systemd systems, that should be fairly straight forward too.

**Comment by chuckbutler** [comment]
@cloudguru-bb  As an update here, i'm going to be adding 2 installation paths which closely mimics the older docker charm.

1) Install the docker.io package from archive. This will be default, as network restricted environments will fail 100% of the time with the current layer implementation. Thats troublesome in many regards. Typically these network restricted envs are running an archive mirror, so they will at least be promised to work if they follow the guides we recommend  for standing up a private cloud.
2) Offer an upstream option so users can grab the latest and greatest from the docker ppa.

Both of which can still continue to use /etc/defaults/docker - which i gleaned from our docker.io package.

The systemd file looks something like the following, and will be included in a future update once we're closer to 16.04 landing, and i've gotten full testing around both install paths

```
[Unit]
Description=Docker Application Container Engine
Documentation=https://docs.docker.com
After=network.target docker.socket
Requires=docker.socket

[Service]
Type=notify
EnvironmentFile=-/etc/default/docker
ExecStart=/usr/bin/docker daemon -H fd:// $DOCKER_OPTS
MountFlags=slave
LimitNOFILE=1048576
LimitNPROC=1048576
LimitCORE=infinity
TimeoutStartSec=0

[Install]
WantedBy=multi-user.target
```

**Comment by simonklb** [comment]
I hit the same issue before finding this issue thread. I pushed the fix I added [here](https://github.com/simonklb/layer-docker/commit/0fe0fe5ef7e13169a797c5cdb0d62931554cbb21).

Do you already have a fix on-going for this or do you want me to create a pull request?

**Comment by cloudguru-bb** [comment]
Thanks for the updates.  When should I start updating the Expeto charms to comply with this thread info ?

Brian

**Comment by simonklb** [comment]
Regarding the question initially asked in by @cloudguru-bb - I think the [login function in charms.docker](https://github.com/juju-solutions/charms.docker/blob/master/charms/docker/docker.py#L64) would be enough if it supported private registries as well.

I'll give it a try!

**Comment by chuckbutler** [comment]
I'm going to re-open this bug pending a full triage of the layer-docker project regarding the SystemD bits referenced above.

**Comment by chuckbutler** [comment]
Hot fix applied @cloudguru-bb 

There will be more work done here this week to close out systemD support and move away from the install script, but the basics are there thanks to @simonklb's contribution.

**Comment by chuckbutler** [comment]
Closed with #56 and prior template/logic updates
